pytorch_tabnet/callbacks.py

```diff
@@ -0,0 +1,222 @@
+        the counter be reset after each improvment
+
+    """
+
+    early_stopping_metric: str
+    is_maximize: bool
+    tol: float = 0.0
+    patience: int = 5
+
+    def __post_init__(self):
+        self.best_epoch = 0
+        self.wait = 0
+        self.best_weights = None
+        self.best_loss = np.inf
+        if self.is_maximize:
+            self.best_loss = -self.best_loss
+        super().__init__()
+
+    def on_epoch_end(self, epoch, logs=None):
+        current_loss = logs.get(self.early_stopping_metric)
+        if current_loss is None:
+            return
+        loss_change = current_loss - self.best_loss
+        max_improved = (self.is_maximize and loss_change > self.tol)
+        min_improved = ((not self.is_maximize) and (-loss_change > self.tol))
+        if max_improved or min_improved:
+            self.best_loss = current_loss
+            self.wait = 1
+            self.best_weights = copy.deepcopy(self.trainer.network.state_dict())
+            self.best_epoch = epoch
+        else:
+            if self.wait >= self.patience:
+                self.trainer._stop_training = True
+            self.wait += 1
+
+    def on_train_end(self, logs=None):
+        self.trainer.best_epoch = self.best_epoch
+        self.trainer.best_cost = self.best_loss
+        self.trainer.network.load_state_dict(self.best_weights)
+        if self.trainer._stop_training:
+            msg = f"\nEarly stopping occured at epoch {self.best_epoch}"
+            msg += (
+                f" with best_{self.early_stopping_metric} = {round(self.best_loss, 5)}"
+            )
+            print(msg)
+
+
+@dataclass
+class History(Callback):
+    """Callback that records events into a `History` object.
+    This callback is automatically applied to
+    every SuperModule.
+
+    Parameters
+    ---------
+    trainer : DeepRecoModel
+        Model class to train
+    verbose : int
+        Print results every verbose iteration
+
+    """
+
+    trainer: Any
+    verbose: int = 1
+
+    def __post_init__(self):
+        super().__init__()
+        self.samples_seen = 0.0
+        self.total_time = 0.0
+
+    def on_train_begin(self, logs=None):
+        self.epoch_metrics = {"loss": []}
+        self.epoch_metrics.update({"lr": []})
+        self.epoch_metrics.update({name: [] for name in self.trainer._metrics_names})
+        self.start_time = logs["start_time"]
+
+    def on_epoch_begin(self, epoch, logs=None):
+        self.batch_metrics = {"loss": 0.0}
+        self.samples_seen = 0.0
+
+    def on_epoch_end(self, epoch, logs=None):
+        for k in self.batch_metrics:
+            self.epoch_metrics[k].append(self.batch_metrics[k])
+        if self.verbose == 0:
+            return
+        if epoch % self.verbose != 0:
+            return
+        msg = f"epoch: {epoch:<4}"
+        for metric_name, metric_value in self.batch_metrics.items():
+            msg += f"|  {metric_name:<5}: {np.round(metric_value, 5):<10}"
```

Comment:
could we reduce the blank spaces between the `|` this would allow the metric to fit in one line more easily.

Also I find it cool to see the learning rate update but it takes a lot of room on the screen while it's not first order info, maybe we should just give people access to the learning rate in the history logs. What do you think?

Reply:
There's also a problem when values get smaller than 1e-5, could we force the scienfitic writting?
![lr_decay_printing_error](https://user-images.githubusercontent.com/12525461/95347575-556d4e00-08bd-11eb-8c84-6964e484d324.png)

Reply:
lr don't appear into the log anymore, just into history

Reply:
all right then